content/documentation/modelHub/content.md

```diff
@@ -8,25 +8,393 @@ weight = 140
+
+Commands are usually handled directly by the Model Service, so they will not be visible to the client of the Model Service.
+
+```ts
+export interface CoffeeModelService {
+  getCoffeeModel(modelUri: string): Promise<CoffeeModelRoot | undefined>;
+
+  unload(modelUri: string): Promise<void>;
+
+  edit(modelUri: string, patch: Operation[]): Promise<PatchResult>;
+
+  createNode(modelUri: string, parent: string | Workflow, args: CreateNodeArgs): Promise<PatchResult>;
+}
+
+/**
+ * This constant is used to register the CoffeeModelService and to retrieve
+ * it from the ModelHub.
+ */ 
+export const COFFEE_SERVICE_KEY = 'coffeeModelService';
+
+// Access and use the model service
+const modelService: CoffeeModelService = modelHub.getModelService<CoffeeModelService>(COFFEE_SERVICE_KEY);
+await modelService.createNode(modelId, '/workflows/0', { type: 'AutomaticTask' });
+```
+
 #### Validating models
 
+Model Contributions may register Validators. These Validators will be invoked whenever the ModelHub is validated:
+
+```ts
+const modelId = 'file:///coffee-editor/examples/workspace/superbrewer3000.coffee';
+const diagnostic = await modelHub.validateModels(modelId);
+```
+
+Several models can be validated at the same time:
+
+```ts
+const modelId1 = 'file:///coffee-editor/examples/workspace/superbrewer2000.coffee';
+const modelId2 = 'file:///coffee-editor/examples/workspace/superbrewer3000.coffee';
+const diagnostic = await modelHub.validateModels(modelId1, modelId2);
+```
+
+Or you can validate all models currently loaded, by omitting the `modelIds` argument:
+
+```ts
+const diagnostic = await modelHub.validateModels();
+```
+
+**Note:** In the latter case, only models currently *loaded* will be validated. Since the model hub relies on lazy-loading to identify existing models, it may ignore some models present in your workspace, if they have never been explicitly loaded beforehand.
+
+The `validateModels` method will validate all requested models, then return the validation results, in the form of a Diagnostic.
+
+If you're only interested in the latest known validation results, but don't want to wait for a full validation cycle, you can use `getValidationState` instead. This method doesn't trigger any validation, but returns the result from the latest validation:
+
+```ts
+const modelId = 'file:///coffee-editor/examples/workspace/superbrewer3000.coffee';
+const currentDiagnostic = modelHub.getValidationState(modelId);
+```
+
 ### Contributing modeling languages
 
-TODO more detailed look at the API for registering a language contribution for the capabilities listed below
+The Model Hub can handle several aspects for each Modeling Language:
+
+- Persistence (Save/Load)
+- Edition (Via Model Services and Commands)
+- Validation
+- Triggers
+
+All of these aspects can be registered using a `ModelServiceContribution`. The only mandatory aspect is Edition, via a Model Service.
```

Comment:
Yeah, it's a bit confusing. For subclasses of AbstractModelServiceContribution, a ModelService is really the only mandatory part (even though it can be empty, it must be defined; it's the only abstract method).

Even persistence is optional, because you could have two ModelServiceContributions for the same type of Model. In that case, only one of them needs to provide a Persistence Contribution. However, both need to provide a Model Service implementation.

From the ModelServiceContribution interface point of view, it's different though: persistence contribution is indeed mandatory, as well as model service ((even though you can provide a No-op for both)